From 16f308738daaccb870418619e96a920605195a0b Mon Sep 17 00:00:00 2001
From: linma <linma@zju.edu.cn>
Date: Mon, 17 May 2021 19:03:11 +0800
Subject: [PATCH] Bluetooth: fix the erroneous flush_work() order

In the cleanup routine for failed initialization of HCI device,
the flush_work(&hdev->rx_work) need to be finished before the
flush_work(&hdev->cmd_work). Otherwise, the hci_rx_work() can
possibly invoke new cmd_work and cause a bug, like double free,
in late processings.

Signed-off-by: linma <linma@zju.edu.cn>
---
 net/bluetooth/hci_core.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/net/bluetooth/hci_core.c b/net/bluetooth/hci_core.c
index fd12f1652bdf..7f8eaf804287 100644
--- a/net/bluetooth/hci_core.c
+++ b/net/bluetooth/hci_core.c
@@ -1610,8 +1610,9 @@ static int hci_dev_do_open(struct hci_dev *hdev)
 	} else {
 		/* Init failed, cleanup */
 		flush_work(&hdev->tx_work);
+		flush_work(&hdev->rx_work);  // rx_work should be flushed before the cmd_work
+		                             // to avoid new cmd_work
 		flush_work(&hdev->cmd_work);
-		flush_work(&hdev->rx_work);
 
 		skb_queue_purge(&hdev->cmd_q);
 		skb_queue_purge(&hdev->rx_q);
-- 
2.30.2


